### PR TITLE
Add async/await support

### PR DESCRIPTION
**Pull request by rossnomann**
### TODO:

- [ ] Add async/await support for:
  - [ ] carapax-access
  - [ ] carapax-i18n
  - [ ] carapax-ratelimit
  - [ ] carapax-session
- [ ] Add socks and rustls support when reqwest 0.10 will be released.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -2,6 +2,7 @@
 members = [
     "carapax",
     "carapax-access",
+    "carapax-codegen",
```
Comment (ark0f) [inline Cargo.toml:5]: Do function handlers need their codegen instead of simple `FnHandler::new`?

File: carapax/Cargo.toml [hunk 2]
```diff
@@ -11,16 +11,15 @@ repository = "https://github.com/tg-rs/tg-rs"
...
-failure = "0.1.5"
-futures = "0.1.23"
-log = "0.4.6"
-regex = "1.0.5"
+async-trait = "0.1.21"
+carapax-codegen = { path = "../carapax-codegen" }
 shellwords = "1.0.0"
 tgbot = { path = "../tgbot" }
-tokio-executor = "0.1.5"
+tokio = "0.2.6"
```
Comment (ark0f) [inline carapax/Cargo.toml:18]: TODO: Set minimal library versions

File: carapax/examples/commands.rs [hunk 3]
```diff
@@ -1,53 +1,50 @@
...
     dotenv().ok();
     env_logger::init();
 
     let token = env::var("TGRS_TOKEN").expect("TGRS_TOKEN is not set");
     let proxy = env::var("TGRS_PROXY").ok();
-
     let mut config = Config::new(token);
     if let Some(proxy) = proxy {
-        config = config.proxy(proxy);
+        config = config.proxy(proxy).expect("Failed to set proxy");
```
Comment (ark0f) [inline carapax/examples/commands.rs:39]: TODO: Use `Into<Option<Config>>`

File: carapax/src/command.rs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+            }
+        )
+    }
+}
+
+impl TryFromUpdate for Command {
+    type Error = CommandError;
+
+    fn try_from_update(update: Update) -> Result<Option<Self>, Self::Error> {
+        Ok(match update.get_message() {
```
Comment (ark0f) [inline carapax/src/command.rs:165]: TODO: Simplify

File: carapax/src/command.rs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[tokio::test]
+    async fn command() {
+        let command = Command::try_from_update(
```
Comment (ark0f) [inline carapax/src/command.rs]: TODO: More convenient testing

File: carapax/src/convert.rs [hunk 6]
```diff
@@ -1,82 +1,104 @@
+use std::{convert::Infallible, error::Error};
 use tgbot::types::{
     CallbackQuery, ChosenInlineResult, InlineQuery, Message, Poll, PreCheckoutQuery, ShippingQuery, Update, UpdateKind,
 };
 
 /// Allows to create an input for a handler from given update
-pub trait FromUpdate: Sized {
+pub trait TryFromUpdate: Sized {
+    /// An error when converting update
+    type Error: Error + Send + Sync;
```
Comment (ark0f) [inline carapax/src/convert.rs:9]: Where are `Send` and `Sync` required?

File: carapax/src/dispatcher.rs [hunk 7]
```diff
@@ -1,264 +1,173 @@
...
-}
-
-pub(crate) struct Dispatcher {
-    api: Api,
-    handlers: Arc<Vec<BoxedHandler>>,
-    error_strategy: ErrorStrategy,
+/// A Telegram Update dispatcher
+pub struct Dispatcher {
+    handlers: Vec<Box<dyn Handler<Input = Update, Output = HandlerResult> + Send>>,
+    context: Arc<Mutex<Context>>,
```
Comment (ark0f) [inline carapax/src/dispatcher.rs]: Get rid of `Mutex`

File: carapax/src/dispatcher.rs [hunk 8]
```diff
@@ -1,264 +1,173 @@
...
-    Abort,
-}
-
-pub(crate) struct Dispatcher {
-    api: Api,
-    handlers: Arc<Vec<BoxedHandler>>,
-    error_strategy: ErrorStrategy,
+/// A Telegram Update dispatcher
+pub struct Dispatcher {
+    handlers: Vec<Box<dyn Handler<Input = Update, Output = HandlerResult> + Send>>,
```
Comment (ark0f) [inline carapax/src/dispatcher.rs]: Does `BoxedHandler` differ?

File: carapax/src/handler.rs [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+use std::{error::Error, fmt};
+use tgbot::types::Update;
+
+/// An update handler
+#[async_trait]
+pub trait Handler {
+    /// An object to handle (update, message, inline query, etc...)
+    ///
+    /// See [TryFromUpdate](trait.TryFromUpdate.html) for more information
+    type Input: TryFromUpdate + Send + Sync;
```
Comment (ark0f) [inline carapax/src/handler.rs:12]: Same as above

File: carapax/src/handler.rs [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+    }
+}
+
+impl From<HandlerError> for HandlerResult {
+    fn from(err: HandlerError) -> Self {
+        HandlerResult::Error(err)
+    }
+}
+
+pub(crate) struct BoxedHandler<H>(H);
```
Comment (ark0f) [inline carapax/src/handler.rs]: It's not boxed anymore

File: tgbot/examples/send_file.rs [hunk 11]
```diff
@@ -22,99 +22,108 @@ struct Handler {
...
-
-        macro_rules! execute {
-            ($method:expr) => {
-                self.api.spawn(self.api.execute($method).then(|x| {
-                    log::info!("method result: {:?}\n", x);
-                    Ok::<(), ()>(())
-                }));
-            };
-        }
+    type Error = Infallible;
```
Comment (ark0f) [inline tgbot/examples/send_file.rs]: `Infallible` with `Api::execute`?

File: tgbot/src/api.rs [hunk 12]
```diff
@@ -42,11 +44,19 @@ impl Config {
...
     /// * socks5://\[user:password\]@host:port
-    pub fn proxy<S: Into<String>>(mut self, proxy: S) -> Self {
-        self.proxy = Some(proxy.into());
-        self
+    pub fn proxy<U: AsRef<str>>(mut self, url: U) -> Result<Self, ParseProxyError> {
+        let raw_url = url.as_ref();
+        let url = Url::parse(raw_url)?;
+        if url.has_authority() {
+            let mut base_url = url.clone();
+            base_url.set_username("").expect("Failed to remove username");
```
Comment (ark0f) [inline tgbot/src/api.rs:53]: TODO: Investigate to better code

File: tgbot/src/api.rs [hunk 13]
```diff
@@ -62,111 +72,267 @@ where
...
     /// Creates a API instance with a given configuration.
-    pub fn new<C: Into<Config>>(config: C) -> Result<Self, Error> {
+    pub fn new<C: Into<Config>>(config: C) -> Result<Self, ApiError> {
         let config = config.into();
+
+        let mut builder = ClientBuilder::new();
+        builder = if let Some(proxy) = config.proxy {
+            builder.proxy(proxy)
+        } else {
+            builder.no_proxy()
```
Comment (ark0f) [inline tgbot/src/api.rs:89]: Do we really need it here?

File: tgbot/src/api.rs [hunk 14]
```diff
@@ -62,111 +72,267 @@ where
...
+        Some(match self {
+            ApiError::BuildClient(err) => err,
+        })
+    }
+}
+
+impl fmt::Display for ApiError {
+    fn fmt(&self, out: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            ApiError::BuildClient(err) => write!(out, "can not build HTTP client: {}", err),
```
Comment (ark0f) [inline tgbot/src/api.rs:180]: Is it better to change to `Failed to build HTTP client`?
@mexus

File: tgbot/src/api.rs [hunk 15]
```diff
@@ -62,111 +72,267 @@ where
...
+            ParseProxyError::Reqwest(err) => err,
+        })
+    }
+}
+
+impl fmt::Display for ParseProxyError {
+    fn fmt(&self, out: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            out,
+            "can not parse proxy URL: {}",
```
Comment (ark0f) [inline tgbot/src/api.rs:207]: Same as above

File: tgbot/src/api.rs [hunk 16]
```diff
@@ -62,111 +72,267 @@ where
...
+        }
+    }
+}
 
-    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-        self.inner.poll()
+impl fmt::Display for DownloadFileError {
+    fn fmt(&self, out: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            DownloadFileError::Reqwest(err) => write!(out, "failed to download file: {}", err),
```
Comment (ark0f) [inline tgbot/src/api.rs:248]: Like here, @mexus

File: tgbot/src/executor/mod.rs [hunk 17]
```diff
@@ -1,11 +0,0 @@
-use crate::request::Request;
-use failure::Error;
-use futures::Future;
-
-mod hyper;
-
-pub(crate) use self::hyper::{default_executor, proxy_executor};
-
-pub(crate) trait Executor: Send + Sync {
-    fn execute(&self, req: Request) -> Box<dyn Future<Item = Vec<u8>, Error = Error> + Send>;
```
Comment (ark0f) [inline tgbot/src/executor/mod.rs:10]: `reqwest` only?

File: tgbot/src/lib.rs [hunk 18]
```diff
@@ -1,23 +1,27 @@
 //! A Telegram Bot API client library
+#![recursion_limit = "256"]
```
Comment (ark0f) [inline tgbot/src/lib.rs:2]: Reason?

File: tgbot/src/longpoll.rs [hunk 19]
```diff
@@ -0,0 +1,196 @@
...
+                for update in updates {
+                    offset = max(offset, update.id);
+                    yield update
+                }
+            }
+        };
+        pin_mut!(s);
+        while let Some(update) = s.next().await {
+            if let Err(err) = self.handler.handle(update).await {
+                error!("Failed to handle update: {}", err);
```
Comment (ark0f) [inline tgbot/src/longpoll.rs]: What do you think about callback here?

File: tgbot/src/longpoll.rs [hunk 20]
```diff
@@ -0,0 +1,196 @@
...
+
+impl<H> LongPoll<H> {
+    /// Creates a new LongPoll
+    ///
+    /// # Arguments
+    ///
+    /// * api - Telegram Bot API Client
+    /// * handler - Updates Handler
+    pub fn new(api: Api, handler: H) -> Self {
+        let (sender, receiver) = channel(1);
```
Comment (ark0f) [inline tgbot/src/longpoll.rs:37]: Why not `oneshot`?

File: tgbot/src/request/form.rs [hunk 21]
```diff
@@ -24,6 +35,52 @@ impl FormValue {
             FormValue::File(ref file) => Some(file),
         }
     }
+
+    async fn into_part(self) -> Result<Part, FormError> {
+        Ok(match self {
+            FormValue::Text(text) => Part::text(text),
+            FormValue::File(file) => match file.kind {
+                InputFileKind::Path(path) => {
+                    let file_name = path.file_name().map(|x| x.to_string_lossy().into_owned());
```
Comment (ark0f) [inline tgbot/src/request/form.rs:44]: Why lossy?

File: tgbot/src/types/message/mod.rs [hunk 22]
```diff
@@ -276,18 +277,34 @@ pub enum EditMessageResult {
     Bool(bool),
 }
 
-#[derive(Debug, failure::Fail, derive_more::From)]
+#[derive(Debug, derive_more::From)]
```
Comment (ark0f) [inline tgbot/src/types/message/mod.rs:280]: I didn't saw `derire_more::From` for errors above

File: tgbot/src/webhook.rs [hunk 23]
```diff
@@ -0,0 +1,165 @@
...
+            path,
+            handler: self.handler.clone(),
+        })
+    }
+}
+
+#[doc(hidden)]
+pub struct WebhookService<H> {
+    path: String,
+    handler: Arc<Mutex<H>>,
```
Comment (ark0f) [inline tgbot/src/webhook.rs:58]: Get rid of `Mutex`

File: tgbot/src/webhook.rs [hunk 24]
```diff
@@ -0,0 +1,165 @@
...
+/// Starts a server for webhook
+///
+/// # Arguments
+///
+/// * address - Bind address
+/// * path - URL path for webhook
+/// * handler - Updates handler
+pub async fn run_server<A, P, H>(address: A, path: P, handler: H) -> Result<(), HyperError>
+where
+    A: Into<SocketAddr>,
```
Comment (ark0f) [inline tgbot/src/webhook.rs:152]: Can `trait ToSocketAddrs` be used here?

File: tgbot/tests/longpoll.rs [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+use std::{
+    convert::Infallible,
+    sync::Arc,
+    time::{Duration, Instant},
+};
+use tgbot::{longpoll::LongPoll, types::Update, Api, Config, UpdateHandler};
+use tokio::{spawn, sync::Mutex, time::delay_for};
+
+struct Handler {
+    updates: Arc<Mutex<Vec<Update>>>,
```
Comment (ark0f) [inline tgbot/tests/longpoll.rs:13]: Get rid of `Mutex`

**Review by ark0f — CHANGES_REQUESTED** [review]
Comments starting with `TODO` can be done after PR merging

**Review by mexus — DISMISSED** [review]
TLDR

**Comment by qezz** [comment]
LGTM

File: tgbot/examples/send_file.rs (hunk 11 above)
Comment (rossnomann) [reply tgbot/examples/send_file.rs]: It panics and never returns an error.

File: carapax/src/convert.rs (hunk 6 above)
Comment (rossnomann) [reply carapax/src/convert.rs:9]: Because we use Mutex

File: carapax/src/dispatcher.rs (hunk 7 above)
Comment (rossnomann) [reply carapax/src/dispatcher.rs]: We can't, because context needs to be mutable

File: carapax/src/handler.rs (hunk 10 above)
Comment (rossnomann) [reply carapax/src/handler.rs]: Your suggestions?

File: tgbot/src/lib.rs (hunk 18 above)
Comment (rossnomann) [reply tgbot/src/lib.rs:2]: async_stream macro

File: tgbot/src/longpoll.rs (hunk 20 above)
Comment (rossnomann) [reply tgbot/src/longpoll.rs:37]: Because we need to shutdown from multiple senders

File: tgbot/src/longpoll.rs (hunk 19 above)
Comment (rossnomann) [reply tgbot/src/longpoll.rs]: I think we can do it later

File: tgbot/src/types/message/mod.rs (hunk 22 above)
Comment (rossnomann) [reply tgbot/src/types/message/mod.rs:280]: Sometimes we need to convert error explicitly

File: tgbot/src/webhook.rs (hunk 23 above)
Comment (rossnomann) [reply tgbot/src/webhook.rs:58]: Handler needs to take &mut self, so we can't

File: tgbot/src/webhook.rs (hunk 24 above)
Comment (rossnomann) [reply tgbot/src/webhook.rs:152]: Server::bind takes SocketAddr

File: tgbot/tests/longpoll.rs (hunk 25 above)
Comment (rossnomann) [reply tgbot/tests/longpoll.rs:13]: Your suggestions?

File: carapax/src/handler.rs (hunk 10 above)
Comment (ark0f) [reply carapax/src/handler.rs]: `WrappingHandler`

File: tgbot/tests/longpoll.rs (hunk 25 above)
Comment (ark0f) [reply tgbot/tests/longpoll.rs:13]: Oh, it is in test. It doesn't matter then

File: tgbot/src/request/form.rs (hunk 21 above)
Comment (rossnomann) [reply tgbot/src/request/form.rs:44]: Ok, I will fix it

File: tgbot/src/api.rs (hunk 13 above)
Comment (rossnomann) [reply tgbot/src/api.rs:89]: If we have not set proxy, we must be sure that proxy is not used

File: tgbot/src/executor/mod.rs (hunk 17 above)
Comment (rossnomann) [reply tgbot/src/executor/mod.rs:10]: Are there any reasons to have support for other HTTP clients?

File: carapax/examples/commands.rs [hunk 26]
```diff
@@ -1,53 +1,46 @@
...
+};
 use dotenv::dotenv;
 use env_logger;
-use futures::Future;
 use log;
 use std::env;
 
-fn handle_start(context: &mut Context, message: Message, _: Vec<String>) -> HandlerFuture {
+#[handler(command = "/start")]
+async fn handle_start(context: &mut Api, command: Command) -> Result<HandlerResult, ExecuteError> {
```
Comment (ark0f) [inline carapax/examples/commands.rs:10]:
```suggestion
async fn handle_start(api: &mut Api, command: Command) -> Result<HandlerResult, ExecuteError> {
```

File: carapax/src/command.rs [hunk 27]
```diff
@@ -0,0 +1,254 @@
...
+};
+use async_trait::async_trait;
+use shellwords::MismatchedQuotes;
+use std::{collections::HashMap, error::Error, fmt, string::FromUtf16Error};
+use tgbot::types::{Message, Update};
+
+/// A simple commands handler
+#[derive(Default)]
+pub struct CommandDispatcher<C> {
+    handlers: HashMap<String, Box<dyn Handler<C, Input = Command, Output = HandlerResult> + Send>>,
```
Comment (ark0f) [inline carapax/src/command.rs:13]:
```rust
pub(crate) type BoxedHandler = Box<dyn Handler<C, Input = Command, Output = HandlerResult> + Send>
```
```suggestion
    handlers: HashMap<String, BoxedHandler>,
```
We can see this type in other code places, can't we?

File: carapax/src/command.rs [hunk 28]
```diff
@@ -0,0 +1,254 @@
...
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn create_command(command: &str) -> Command {
+        let len = command.split(" ").next().unwrap().len();
```
Comment (ark0f) [inline carapax/src/command.rs:195]:
```suggestion
        let len = command.split_whitespace().next().unwrap().len();
```

File: carapax/src/lib.rs [hunk 29]
```diff
@@ -1,19 +1,67 @@
...
-pub mod context;
+/// Mark an async function as update handler
+///
+/// If called without arguments it simply will implement [Handler](trait.Handler.html) for given function.
+/// Example:
+/// ```
+/// use carapax::{types::Update, handler};
+///
+/// #[handler]
+/// async fn handle_update(context: &mut (), update: Update) {
```
Comment (ark0f) [inline carapax/src/lib.rs:29]: Make it optional
```suggestion
/// async fn handle_update(update: Update) {
```

File: tgbot/src/request/form.rs (hunk 21 above)
Comment (ark0f) [reply tgbot/src/request/form.rs:44]: Still not fixed

**Review by ark0f — CHANGES_REQUESTED** [review]
